pip/pip-447.md

```diff
@@ -0,0 +1,230 @@
+
+## Forward Compatibility
+
+Prometheus Systems: If a Pulsar broker with this feature enabled sends metrics with custom metric labels to an older Prometheus server or a monitoring system not expecting these additional labels, those systems will typically ignore the extra labels without issue.
+
+Future Enhancements: Future Pulsar versions could extend this feature, for example, by allowing more dynamic management of allowedCustomMetricLabelKeys if deemed safe and necessary.
+
+OpenTelemetry Alignment: The key-value structure of custom metric labels aligns well with OpenTelemetry attributes, ensuring that this feature remains relevant and compatible with Pulsar's evolving metrics infrastructure.
+
+# Testing Plan
+
+A comprehensive testing strategy will be required:
+
+Unit Tests: For new logic in pulsar-admin, REST API handlers, policy management, and metrics generation.
+
+Integration Tests:
+
+Verify correct setting, getting, and removing of custom metric labels via admin tools and REST APIs.
+
+Test validation logic for allowed keys, max labels per topic, and value length.
+
+Ensure correct propagation of custom metric label policy changes via system topics and updates to broker policy caches.
+
+Verify that Prometheus metrics output accurately includes the custom metric labels for relevant topics and does not include them when the feature is disabled or labels are not set.
+
+End-to-End Tests: Simulate a Pulsar cluster environment, set custom metric labels on topics, and scrape metrics using a Prometheus instance to confirm labels appear correctly and can be queried.
+
+Performance Tests: Assess any potential performance impact on brokers, particularly concerning policy updates and metrics generation, especially in environments with a large number of topics.
+
+# Documentation Plan
+
+The official Apache Pulsar documentation will be updated to include:
+
+Concepts Section: Explanation of the custom topic metric labels feature, its purpose, and how it helps with monitoring and alerting.
+
+Administrator Guide:
+
+Instructions on how to enable and configure the feature in broker.conf (e.g., exposeCustomTopicMetricLabelsEnabled, allowedCustomMetricLabelKeys, and other limits).
+
+Best practices for defining allowedCustomMetricLabelKeys with cardinality management in mind.
+
+User Guide / pulsar-admin Reference:
+
+Detailed syntax and examples for the new pulsar-admin topics set/get/remove-custom-metric-labels commands.
+
+Guidance on choosing appropriate values for predefined keys to manage cardinality.
+
+REST API Reference: Documentation for the new REST API endpoints.
+
+Monitoring Section: Notes on how these custom metric labels appear in Prometheus and how they can be used in PromQL queries, along with reminders about cardinality considerations for the Prometheus system.
+
+# Alternatives
+
+Briefly, two other approaches were considered and rejected:
```

Comment:
Were topic metadata (properties) added in [PIP-110](https://github.com/apache/pulsar/blob/master/pip/pip-110.md) considered as the data source for labels?